homeassistant/components/sunweg/sensor.py

```diff
@@ -154,25 +123,44 @@ def __init__(
         )
 
         self._attr_device_info = DeviceInfo(
-            identifiers={(DOMAIN, str(probe.plant_id))},
+            identifiers={(DOMAIN, str(coordinator.plant_id))},
             manufacturer="SunWEG",
-            name=name,
+            name=coordinator.plant_name,
         )
 
-    def update(self) -> None:
-        """Get the latest data from the Sun WEG API and updates the state."""
-        self.probe.update()
-        (
-            self._attr_native_value,
-            self._attr_native_unit_of_measurement,
-        ) = self.probe.get_data(
-            api_variable_key=self.entity_description.api_variable_key,
-            api_variable_unit=self.entity_description.api_variable_unit,
-            deep_name=self.deep_name,
-            device_type=self.device_type,
-            inverter_id=self.inverter_id,
-            name=self.entity_description.name,
-            native_unit_of_measurement=self.native_unit_of_measurement,
-            never_resets=self.entity_description.never_resets,
-            previous_value_drop_threshold=self.entity_description.previous_value_drop_threshold,
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle data update."""
+        previous_value = self.native_value
+        value: StateType | datetime.datetime = self.coordinator.get_api_value(
+            self.entity_description.api_variable_key,
+            self.device_type,
+            self.inverter_id,
+            self.deep_name,
         )
+        previous_unit_of_measurement: str | None = self.native_unit_of_measurement
```

Comment:
This seems very strange. We should set the UoM once during init and not change it.

Reply:
Hey @gjohansson-ST, I asked this question when I did my review and [the answer](https://github.com/home-assistant/core/pull/115356#discussion_r1567505336) seemed to suggest that the plant could be offline at startup but sill return data.

Reply:
Sure, but it doesn't give any answer to how an UoM can change for the sensors. It doesn't make sense.

Reply:
This happens because the API returns it as a string so they can change the unit anytime. Do you see a better way to deal with it?

Reply:
Can you explain a bit more? The UoM can be `kWh` and then the user can just change it to `mm` if they want?
How does this affect the value (from the api side)?

Reply:
The API can return "4.12 MWh" or "4,123.2 kWh". I'm parsing the string to provide the correct number to the integration. And when the plant is offline and the day changes, the daily generation will return null. And this is threated as None in the unit part.

Reply:
Ok, it would be better to try to handle that in the upstream library to always return the same UoM to HA.
Anyhow maybe we can put a comment into the code about this so for next reviewer it's more obvious than asking the same question again?

Reply:
I think it would be nice to normalize this in the upstream library indeed.

Reply:
I think it's also interesting to see, if the device is offline, shouldn't we render the device as unavailable?